### DISCUSSION TITLE
Highlighting functional groups

### DISCUSSION THREAD

**Discussion by Keytoll** [rdkit/rdkit#7614] [Development]
Hello, 

I've been trying to reliably highlight many functional groups with a different colour assigned to each group in any molecule so that you could have kind of like a rainbow heatmap to see different functional groups that also highlight where there is overlap. 

Because substructure detection/functional group detection is such a core part of chemistry I thought that this would be relatively easy, but am having trouble when highlighting, often there are false positives etc, is there already a worked solution to this? Would the functionalgroups.txt, hierarchy etc be useful for this or? I'm not finding a lot of documentation, any insights would be very much appreciated.

 I think its relatively easy finding out if such a functional group exists, but when it comes to highlighting and knowing when it starts and when it ends for many functional groups there is added complexity in which I am not experienced enough to handle right now.

Thank you.

**Comment by HelloJocelynLu** [comment]
I developed a fragmentation method that can help [highlight functional groups](https://github.com/HelloJocelynLu/EFGs/blob/master/Examples/Color%20EFGs%20by%20their%20molecular%20weights.ipynb), inspired by [Peter Ertl](https://jcheminf.biomedcentral.com/articles/10.1186/s13321-017-0225-z)'s work:
https://github.com/HelloJocelynLu/EFGs.
This task may not as easy as you think, as definitions of functional groups can vary. A more flexible option: You can also define a list of functional groups of interest using [SMARTS](https://www.daylight.com/dayhtml_tutorials/languages/smarts/smarts_examples.html) to [highlight](https://www.rdkit.org/docs/GettingStartedInPython.html#drawing-molecules) the matched components.

**Reply by Keytoll** [reply]
Very cool, I've had a try using your repo and highlight example, and for my use case I want to add explicit hydrogens and highlight them accordingly so I did .AddHs to the mol and added the TreatHs='include' option. However, I'm getting this error: Check your molecule!!! list element larger than allowed value.

Do you have any idea how to get it to work for this use case?

EDIT: don't worry, problem on my end.

**Reply by HelloJocelynLu** [reply]
No worries at all! Glad you were able to figure it out. You are right, that drawing function is a bit old and I might need to revisit it someday...
